### DISCUSSION TITLE
Error reading the file

### DISCUSSION THREAD

**Discussion by mpetris3** [simdjson/simdjson#1487] [Q&A]
Hi everyone,

I start by saying that I am not very proficient in c++ and it's the first time I need to parse json files.   
I would like to use simdjson for a project where I need to parse quite large json files 2, 3, 4.. gb.
I tried to parse these files both with DOM and ondemand, but I keep getting the "Error reading file" exception.
The exception is thrown respectively at lines 
simdjson::dom::element tweets = parser.load(filename); 
or
ondemand::document doc = parser.iterate(json);

I manage to parse smaller files with simdjson and I succeded in parsing those large ones by means of another library.
I may miss something obvious, but can you please help me with this issue? 
Thanks.

**Comment by lemire** [comment]
Would you kindly run the following C program on your data... and report back?

```C
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char **argv) {
  if (argc < 2) {
    perror("provide a file argument");
  }
  FILE *pFile = fopen(argv[1], "rb");
  if (pFile == NULL)
    perror("Error opening file");
  else {
    if (fseek(pFile, 0, SEEK_END) < 0) {
      perror("cannot seek the end of the file");
      if (fclose(pFile) != 0) {
        perror("can't close");
      }
      return EXIT_FAILURE;
    } else {
      long size = ftell(pFile);
      printf("Size of %s: %ld bytes.\n", argv[1], size);
      rewind(pFile);
      char *buffer = (char *)malloc(size + 32);
      if (buffer == NULL) {
        perror("can't allocate");
      } else {
        size_t bytes_read = fread(buffer, 1, size, pFile);
        printf("Read: %zu bytes.\n", bytes_read);
        free(buffer);
      }
      if (fclose(pFile) != 0) {
        perror("can't close");
        return EXIT_FAILURE;
      }
    }
  }
  printf("Completed succesfully\n");
  return EXIT_SUCCESS;
}
```

Gist: https://gist.github.com/lemire/b2ef735747a4b1acebcfb47cb81dbe21

As far as I know, it is an entirely standard way to load a file using standard C.

If it works, you are basically done because you can just pass `buffer` to simdjson.

If it works, would then run the following C++ program (requires C++11 and simdjson):

```C++
#include <iostream>
#include "simdjson.h"
#include "simdjson.cpp"

int main(int argc, char *argv[]) {
  if(argc < 2) {
    std::cerr << "Please specify at least one file name. " << std::endl;
    return EXIT_FAILURE;
  }
  const char * filename = argv[1];
  simdjson::padded_string p = simdjson::padded_string::load(filename);
  std::cout << "read " << p.size() << std::endl;
  return EXIT_SUCCESS;
}
```

It should report back the file size.

Here is the load function in question...

```C++

inline simdjson_result<padded_string> padded_string::load(const std::string &filename) noexcept {
  // Open the file
  SIMDJSON_PUSH_DISABLE_WARNINGS
  SIMDJSON_DISABLE_DEPRECATED_WARNING // Disable CRT_SECURE warning on MSVC: manually verified this is safe
  std::FILE *fp = std::fopen(filename.c_str(), "rb");
  SIMDJSON_POP_DISABLE_WARNINGS

  if (fp == nullptr) {
    return IO_ERROR;
  }

  // Get the file size
  if(std::fseek(fp, 0, SEEK_END) < 0) {
    std::fclose(fp);
    return IO_ERROR;
  }
  long llen = std::ftell(fp);
  if((llen < 0) || (llen == LONG_MAX)) {
    std::fclose(fp);
    return IO_ERROR;
  }

  // Allocate the padded_string
  size_t len = static_cast<size_t>(llen);
  padded_string s(len);
  if (s.data() == nullptr) {
    std::fclose(fp);
    return MEMALLOC;
  }

  // Read the padded_string
  std::rewind(fp);
  size_t bytes_read = std::fread(s.data(), 1, len, fp);
  if (std::fclose(fp) != 0 || bytes_read != len) {
    return IO_ERROR;
  }

  return s;
}
```

As you can see, it follows closely the C program above. If the C program works but the load function does not, then we might have an issue to investigate.

**Reply by mpetris3** [reply]
Thank you for the fast reply. 
I successfully run the .c program

Size of name.json: 2726410886 bytes.
Read: 2726410886 bytes.
Completed succesfully

However, when I run the c++ program I get:
filename name.json
Simdjson exception caught: Error reading the file.

(I add a try catch to your piece of code to get it).

**Reply by lemire** [reply]
As you may notice, I wrote the C program by copying and pasting the simdjson code. It really does the same thing... Using either a debugger or logging, can you determine what is happening? I.e., where do the load function and the C program differ.

**Reply by mpetris3** [reply]
Yes. 
So the exception is caught here:
long len = std::ftell(fp);
  if((len < 0) || (len == LONG_MAX)) {
    std::fclose(fp);
    return IO_ERROR;
  }

Specifically, when I consider an input that I successfully parse, len is equal to the size of it.
On the other hand, for the inputs I cannot parse, len is set to -1 and the error I get with perror is 
The following error occurred: Invalid argument.

**Reply by lemire** [reply]
Can you elaborate? You ran the C program... It printed out...

```
Size of name.json: 2726410886 bytes.
Read: 2726410886 bytes.
Completed succesfully
```

This means that the line...

```C++
long size = ftell(pFile);
```

returned 2726410886.

You are now saying that this line within simdjson...

```C++
long len = std::ftell(fp);
```

returns -1.

How can `ftell` return -1 when called by simdjson and 2726410886 when called by a C program? It is the same function and presumably the same file.

**Reply by mpetris3** [reply]
The input files are the same. 
For the C program, I copy paste the code you sent me, and run it through CMD.
For the C++ program (second code you sent me), I have a visual studio project where I work.

I think it's more a problem on my side because even running only the following lines:

std::FILE* fp = std::fopen(argv[1], "rb");
// Get the file size
if (std::fseek(fp, 0, SEEK_END) < 0) {
       std::fclose(fp);
}
long llen = std::ftell(fp);

the program returns -1.
The only way I found to get the correct size of the input is replacing fseek and ftell with, respectively, _fseeki64 and _ftelli64.
But I do not know why the c program works with fseek, ftell.

**Reply by lemire** [reply]
Ah. Please see the PR I wrote. Visual Studio’s ftell is limited to 32-bit sizes...  my PR will extend support for 64-bit inputs under Visual Studio. Meanwhile you can load the files to memory yourself. You are not required to use simdjson to load the data from files.

**Reply by mpetris3** [reply]
Ok. Thank you. I will do as you say.
One last question, if I understood correctly, there is no upper bound on the size of the files that can be parsed with simdjson (once loaded as you said), correct?

**Reply by lemire** [reply]
The parser has a limited maximal capacity which you may control (see API) but we forbid capacities above 4 GB.

**Reply by lemire** [reply]
But to answer your question, if the input does not exceed 4 GB, then it should work fine. Of course, dealing with very large files will require much RAM. You should also configure your system accordingly (e.g., with huge pages) or else you may have suboptimal performance.

**Reply by mpetris3** [reply]
ok! thank you for the answers and time.

**Comment by lemire** [comment]
Importantly, you do not have to use our load functions, you can load the data yourself, using any means you see fit and just pass a pointer. Just make sure to overallocate by SIMDJSON_PADDING bytes and you are set. Our file loading functions are just for convenience.

**Comment by lemire** [comment]
Please see PR https://github.com/simdjson/simdjson/pull/1488
